Add CI

Running ``vulkaninfo`` on CI fails with the error
```
Cannot create Vulkan instance.
This problem is often caused by a faulty installation of the Vulkan driver or attempting to use a GPU that does not support Vulkan.
/build/vulkan-tools-KEbD_A/vulkan-tools-1.2.131.1+dfsg1/vulkaninfo/vulkaninfo.h:371: failed with ERROR_INCOMPATIBLE_DRIVER
Error: Process completed with exit code 255.
```
See https://github.com/ami-iit/yarp-device-openxrheadset/runs/6303508564?check_suite_focus=true

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -123,6 +123,7 @@ jobs:
     - name: Test
       shell: bash -l {0}
       run: |
+        export PATH=$PATH:${GITHUB_WORKSPACE}/install/bin
         XRT_COMPOSITOR_FORCE_XCB=1 monado-service & 
         sleep 10
 

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -65,7 +65,7 @@ jobs:
         sudo apt install wget unzip ninja-build curl patch python3 pkg-config libx11-dev glslang-tools libglvnd-dev ca-certificates libusb-1.0-0-dev libudev-dev libhidapi-dev libwayland-dev libuvc-dev libavcodec-dev libopencv-dev libv4l-dev libcjson-dev libsdl2-dev libegl1-mesa-dev 
         
         #Xvfb, just to have a dummy screen
-        sudo apt install xvfb vulkan-utils vulkan-swrast
+        sudo apt install xvfb vulkan-utils
    
     - name: Vulkan extensions
       shell: bash -l {0}

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -78,18 +78,9 @@ jobs:
         cd ${GITHUB_WORKSPACE}
         git clone https://gitlab.freedesktop.org/monado/monado.git --depth 1 --branch ${MONADO_TAG}
         cd monado && mkdir -p build && cd build
-        cmake -G Ninja -DCMAKE_INSTALL_PREFIX=${GITHUB_WORKSPACE}/install -DCMAKE_BUILD_TYPE=Release -DXRT_HAVE_VULKAN:BOOL=OFF ..
+        cmake -G Ninja -DCMAKE_INSTALL_PREFIX=${GITHUB_WORKSPACE}/install -DCMAKE_BUILD_TYPE=Release -DXRT_HAVE_VULKAN:BOOL=OFF -DXRT_FEATURE_SERVICE:BOOL=OFF..
         ninja install
         
-    - name: Test
-      shell: bash -l {0}
-      run: |
-        export PATH=$PATH:${GITHUB_WORKSPACE}/install/bin
-        
-        # Run monado-service
-        XRT_COMPOSITOR_FORCE_XCB=1 xvfb-run monado-service & 
-        sleep 10
-
     - name: Source-based YARP installation
       shell: bash -l {0}
       run: |
@@ -131,5 +122,16 @@ jobs:
       run: |
         cd build
         make install
+        
+    - name: Test
+      shell: bash -l {0}
+      run: |
+        export PATH=$PATH:${GITHUB_WORKSPACE}/install/bin
+        
+        yarpserver --write &
+        
+        # Run monado-service
+        XRT_COMPOSITOR_FORCE_XCB=1 XR_RUNTIME_JSON=${GITHUB_WORKSPACE}/install/share/openxr/1/openxr_monado.json xvfb-run ./build/bin/openxrheadset_test
+        sleep 10
     
 

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -1,0 +1,100 @@
+name: Continuous Integration with tests on Ubuntu using apt dependencies
+
+on:
+  push:
+  pull_request:
+  schedule:
+  # * is a special character in YAML so you have to quote this string
+  # Execute a "nightly" build at 2 AM UTC
+  - cron:  '0 2 * * *'
+  
+env:
+  YARP_TAG: v3.6.0
+  OPENXR_TAG: main
+  MONADO_TAG: main
+
+jobs:
+  build:
+    name: '[${{ matrix.os }}@${{ matrix.build_type }}]'
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        build_type: [Release, Debug]
+        os: [ubuntu-latest]
+
+    steps:
+    - uses: actions/checkout@master
+
+
+    - name: Display environment variables
+      shell: bash
+      run: env
+
+    # Remove apt repos that are known to break from time to time
+    # See https://github.com/actions/virtual-environments/issues/323
+    - name: Remove broken apt repos
+      run: |
+        for apt_file in `grep -lr microsoft /etc/apt/sources.list.d/`; do sudo rm $apt_file; done
+
+    # ============
+    # DEPENDENCIES
+    # ============
+
+    - name: System dependencies
+      run: |
+        sudo apt update
+        sudo apt install build-essential git wget unzip cmake ninja-build libeigen3-dev curl patch python3 pkg-config libxcb-dri2-0-dev libxcb-glx0-dev libxcb-icccm4-dev libxcb-keysyms1-dev libx11-dev libx11-xcb-dev libxxf86vm-dev libxrandr-dev libxcb-randr0-dev libvulkan-dev glslang-tools libglvnd-dev libgl1-mesa-dev ca-certificates libusb-1.0-0-dev libudev-dev libhidapi-dev libwayland-dev libuvc-dev libavcodec-dev libopencv-dev libv4l-dev libcjson-dev libsdl2-dev libegl1-mesa-dev mesa-common-dev libace-dev
+        
+    - name: Source-based OpenXR installation
+      shell: bash -l {0}
+      run: |
+        cd ${GITHUB_WORKSPACE}
+        git clone https://github.com/KhronosGroup/OpenXR-SDK.git --depth 1 --branch ${OPENXR_TAG}
+        cd OpenXR-SDK && mkdir -p build && cd build
+        cmake -DCMAKE_INSTALL_PREFIX=${GITHUB_WORKSPACE}/install -DCMAKE_BUILD_TYPE=Release ..
+        make install
+        
+    - name: Source-based monado installation
+      shell: bash -l {0}
+      run: |
+        cd ${GITHUB_WORKSPACE}
+        git clone https://gitlab.freedesktop.org/monado/monado.git --depth 1 --branch ${MONADO_TAG}
+        cd monado && mkdir -p build && cd build
+        cmake -G Ninja -DCMAKE_INSTALL_PREFIX=${GITHUB_WORKSPACE}/install -DCMAKE_BUILD_TYPE=Release ..
+        ninja install
+
+    - name: Source-based YARP installation
+      shell: bash -l {0}
+      run: |
+        cd ${GITHUB_WORKSPACE}
+        git clone https://github.com/robotology/yarp.git --depth 1 --branch ${YARP_TAG}
+        cd yarp && mkdir -p build && cd build
+        cmake -DCMAKE_PREFIX_PATH=${GITHUB_WORKSPACE}/install -DCMAKE_INSTALL_PREFIX=${GITHUB_WORKSPACE}/install -DCMAKE_BUILD_TYPE=Release ..
+        make install
+
+
+
+    # ===================
+    # CMAKE-BASED PROJECT
+    # ===================
+
+    - name: Configure
+      shell: bash -l {0}
+      run: |
+        mkdir -p build
+        cd build
+        cmake -DCMAKE_PREFIX_PATH=${GITHUB_WORKSPACE}/install -DCMAKE_BUILD_TYPE=${{ matrix.build_type }} \
+              -DCMAKE_INSTALL_PREFIX=${GITHUB_WORKSPACE}/install -DBUILD_TESTING:BOOL=ON ..
+
+    - name: Build
+      shell: bash -l {0}
+      run: |
+        cd build
+        make
+
+    - name: Install
+      shell: bash -l {0}
+      run: |
+        cd build
+        make install
+

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -58,6 +58,9 @@ jobs:
         
         # Monado additional dependencies (just for testing)
         sudo apt install wget unzip ninja-build curl patch python3 pkg-config libx11-dev glslang-tools libglvnd-dev ca-certificates libusb-1.0-0-dev libudev-dev libhidapi-dev libwayland-dev libuvc-dev libavcodec-dev libopencv-dev libv4l-dev libcjson-dev libsdl2-dev libegl1-mesa-dev 
+        
+        #Xvfb, just to have a dummy screen
+        sudo apt install xvfb
 
     
     - name: Source-based OpenXR installation
@@ -123,6 +126,8 @@ jobs:
     - name: Test
       shell: bash -l {0}
       run: |
+        export DISPLAY=:99
+        sudo Xvfb -ac :99 -screen 0 1280x1024x24 > /dev/null 2>&1 &
         export PATH=$PATH:${GITHUB_WORKSPACE}/install/bin
         XRT_COMPOSITOR_FORCE_XCB=1 monado-service & 
         sleep 10

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -44,6 +44,11 @@ jobs:
     - name: System dependencies
       run: |
         sudo apt update
+        
+        # PPA for updated graphics drivers
+        sudo add-apt-repository ppa:oibaf/graphics-drivers
+        sudo apt-get update
+
         # Common dependencies
         sudo apt install build-essential git cmake
         
@@ -61,6 +66,13 @@ jobs:
         
         #Xvfb, just to have a dummy screen
         sudo apt install xvfb vulkan-utils
+        
+    -name: Prepare Vulkan SDK
+      uses: humbletim/setup-vulkan-sdk@v1.2.0
+      with:
+        vulkan-query-version: 1.3.204.0
+        vulkan-components: Vulkan-Headers, Vulkan-Loader
+        vulkan-use-cache: true
         
     - name: Vulkan extensions
       shell: bash -l {0}
@@ -127,15 +139,7 @@ jobs:
       run: |
         cd build
         make install
-        
-    #Copied from https://github.com/godotengine/godot/pull/47414
-    - name: Download SwiftShader
-      run: |
-        wget https://github.com/godotengine/regression-test-project/releases/download/_ci-deps/swiftshader-ubuntu20.04-20211002.zip
-        unzip swiftshader-ubuntu20.04-20211002.zip
-        curr="$(pwd)/libvk_swiftshader.so"
-        sed -i "s|PATH_TO_CHANGE|$curr|" vk_swiftshader_icd.json
-        
+             
     - name: Test
       shell: bash -l {0}
       run: |
@@ -144,6 +148,6 @@ jobs:
         yarpserver --write &
         
         # Run monado-service
-        VK_ICD_FILENAMES=$(pwd)/vk_swiftshader_icd.json DRI_PRIME=0  XRT_COMPOSITOR_FORCE_XCB=1 XR_RUNTIME_JSON=${GITHUB_WORKSPACE}/install/share/openxr/1/openxr_monado.json xvfb-run ./build/bin/openxrheadset_test
+        LIBGL_ALWAYS_SOFTWARE=true XRT_COMPOSITOR_FORCE_XCB=1 XR_RUNTIME_JSON=${GITHUB_WORKSPACE}/install/share/openxr/1/openxr_monado.json xvfb-run ./build/bin/openxrheadset_test
     
 

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -66,14 +66,7 @@ jobs:
         
         #Xvfb, just to have a dummy screen
         sudo apt install xvfb vulkan-utils
-        
-    -name: Prepare Vulkan SDK
-      uses: humbletim/setup-vulkan-sdk@v1.2.0
-      with:
-        vulkan-query-version: 1.3.204.0
-        vulkan-components: Vulkan-Headers, Vulkan-Loader
-        vulkan-use-cache: true
-        
+   
     - name: Vulkan extensions
       shell: bash -l {0}
       run: |

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -9,6 +9,7 @@ on:
   - cron:  '0 2 * * *'
   
 env:
+  YCM_TAG: v0.14.0
   YARP_TAG: v3.6.0
   OPENXR_TAG: main
   MONADO_TAG: main
@@ -66,6 +67,14 @@ jobs:
     - name: Source-based YARP installation
       shell: bash -l {0}
       run: |
+      
+        # YCM
+        cd ${GITHUB_WORKSPACE}
+        git clone -b ${YCM_TAG} https://github.com/robotology/ycm
+        cd ycm && mkdir -p build && cd build
+        cmake -DCMAKE_PREFIX_PATH=${GITHUB_WORKSPACE}/install -DCMAKE_INSTALL_PREFIX=${GITHUB_WORKSPACE}/install -DCMAKE_BUILD_TYPE=Release ..
+        make install
+        
         cd ${GITHUB_WORKSPACE}
         git clone https://github.com/robotology/yarp.git --depth 1 --branch ${YARP_TAG}
         cd yarp && mkdir -p build && cd build

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -60,7 +60,12 @@ jobs:
         sudo apt install wget unzip ninja-build curl patch python3 pkg-config libx11-dev glslang-tools libglvnd-dev ca-certificates libusb-1.0-0-dev libudev-dev libhidapi-dev libwayland-dev libuvc-dev libavcodec-dev libopencv-dev libv4l-dev libcjson-dev libsdl2-dev libegl1-mesa-dev 
         
         #Xvfb, just to have a dummy screen
-        sudo apt install xvfb
+        sudo apt install xvfb vulkan-utils
+        
+    - name: Vulkan extensions
+      shell: bash -l {0}
+      run: |
+        vulkaninfo
 
     
     - name: Source-based OpenXR installation

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -44,11 +44,22 @@ jobs:
     - name: System dependencies
       run: |
         sudo apt update
-        # OpenXR Monado and Yarp
-        sudo apt install build-essential git wget unzip cmake ninja-build libeigen3-dev curl patch python3 pkg-config libxcb-dri2-0-dev libxcb-glx0-dev libxcb-icccm4-dev libxcb-keysyms1-dev libx11-dev libx11-xcb-dev libxxf86vm-dev libxrandr-dev libxcb-randr0-dev libvulkan-dev glslang-tools libglvnd-dev libgl1-mesa-dev ca-certificates libusb-1.0-0-dev libudev-dev libhidapi-dev libwayland-dev libuvc-dev libavcodec-dev libopencv-dev libv4l-dev libcjson-dev libsdl2-dev libegl1-mesa-dev mesa-common-dev libace-dev
-        # Additional dependencies
-        sudo apt install libglm-dev libglfw3-dev libglew-dev
+        # Common dependencies
+        sudo apt install build-essential git cmake
         
+        # OpenXR dependencies
+        sudo apt install libgl1-mesa-dev libvulkan-dev libx11-xcb-dev libxcb-dri2-0-dev libxcb-glx0-dev libxcb-icccm4-dev libxcb-keysyms1-dev libxcb-randr0-dev libxrandr-dev libxxf86vm-dev mesa-common-dev
+        
+        # Yarp dependencies
+        sudo apt install libace-dev
+        
+        # yarp-device-openxrheadset dependencies
+        sudo apt install libeigen3-dev libglm-dev libglfw3-dev libglew-dev
+        
+        # Monado additional dependencies (just for testing)
+        sudo apt install wget unzip ninja-build curl patch python3 pkg-config libx11-dev glslang-tools libglvnd-dev ca-certificates libusb-1.0-0-dev libudev-dev libhidapi-dev libwayland-dev libuvc-dev libavcodec-dev libopencv-dev libv4l-dev libcjson-dev libsdl2-dev libegl1-mesa-dev 
+
+    
     - name: Source-based OpenXR installation
       shell: bash -l {0}
       run: |
@@ -85,7 +96,6 @@ jobs:
         make install
 
 
-
     # ===================
     # CMAKE-BASED PROJECT
     # ===================
@@ -109,4 +119,10 @@ jobs:
       run: |
         cd build
         make install
+        
+    - name: Test
+      shell: bash -l {0}
+      run: |
+        XRT_COMPOSITOR_FORCE_XCB=1 monado-service & 
+        sleep 10
 

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -78,7 +78,7 @@ jobs:
         cd ${GITHUB_WORKSPACE}
         git clone https://gitlab.freedesktop.org/monado/monado.git --depth 1 --branch ${MONADO_TAG}
         cd monado && mkdir -p build && cd build
-        cmake -G Ninja -DCMAKE_INSTALL_PREFIX=${GITHUB_WORKSPACE}/install -DCMAKE_BUILD_TYPE=Release -DXRT_HAVE_VULKAN:BOOL=OFF -DXRT_FEATURE_SERVICE:BOOL=OFF..
+        cmake -G Ninja -DCMAKE_INSTALL_PREFIX=${GITHUB_WORKSPACE}/install -DCMAKE_BUILD_TYPE=Release -DXRT_HAVE_VULKAN:BOOL=OFF -DXRT_FEATURE_SERVICE:BOOL=OFF ..
         ninja install
         
     - name: Source-based YARP installation

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -78,8 +78,17 @@ jobs:
         cd ${GITHUB_WORKSPACE}
         git clone https://gitlab.freedesktop.org/monado/monado.git --depth 1 --branch ${MONADO_TAG}
         cd monado && mkdir -p build && cd build
-        cmake -G Ninja -DCMAKE_INSTALL_PREFIX=${GITHUB_WORKSPACE}/install -DCMAKE_BUILD_TYPE=Release ..
+        cmake -G Ninja -DCMAKE_INSTALL_PREFIX=${GITHUB_WORKSPACE}/install -DCMAKE_BUILD_TYPE=Release -DXRT_HAVE_VULKAN:BOOL=OFF ..
         ninja install
+        
+    - name: Test
+      shell: bash -l {0}
+      run: |
+        export PATH=$PATH:${GITHUB_WORKSPACE}/install/bin
+        
+        # Run monado-service
+        XRT_COMPOSITOR_FORCE_XCB=1 xvfb-run monado-service & 
+        sleep 10
 
     - name: Source-based YARP installation
       shell: bash -l {0}
@@ -123,21 +132,4 @@ jobs:
         cd build
         make install
     
-    #Following https://github.com/godotengine/godot/pull/47414    
-    - name: Download SwiftShader
-      run: |
-        wget https://github.com/qarmin/gtk_library_store/releases/download/3.24.0/swiftshader2.zip
-        unzip swiftshader2.zip
-        rm swiftshader2.zip
-        curr="$(pwd)/libvk_swiftshader.so"
-        sed -i "s|PATH_TO_CHANGE|$curr|" vk_swiftshader_icd.json
-        
-    - name: Test
-      shell: bash -l {0}
-      run: |
-        export PATH=$PATH:${GITHUB_WORKSPACE}/install/bin
-        
-        # Run monado-service
-        XRT_COMPOSITOR_FORCE_XCB=1 VK_ICD_FILENAMES=$(pwd)/vk_swiftshader_icd.json DRI_PRIME=0 xvfb-run monado-service & 
-        sleep 10
 

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -135,6 +135,7 @@ jobs:
              
     - name: Test
       shell: bash -l {0}
+      if : false #Disabling until https://gitlab.freedesktop.org/monado/monado/-/issues/170 is fixed
       run: |
         export PATH=$PATH:${GITHUB_WORKSPACE}/install/bin
         

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -65,7 +65,7 @@ jobs:
         sudo apt install wget unzip ninja-build curl patch python3 pkg-config libx11-dev glslang-tools libglvnd-dev ca-certificates libusb-1.0-0-dev libudev-dev libhidapi-dev libwayland-dev libuvc-dev libavcodec-dev libopencv-dev libv4l-dev libcjson-dev libsdl2-dev libegl1-mesa-dev 
         
         #Xvfb, just to have a dummy screen
-        sudo apt install xvfb vulkan-utils
+        sudo apt install xvfb vulkan-utils mesa-vulkan-drivers
    
     - name: Vulkan extensions
       shell: bash -l {0}

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -125,13 +125,12 @@ jobs:
     
     #Following https://github.com/godotengine/godot/pull/47414    
     - name: Download SwiftShader
-        if: ${{ matrix.tests }}
-        run: |
-          wget https://github.com/qarmin/gtk_library_store/releases/download/3.24.0/swiftshader2.zip
-          unzip swiftshader2.zip
-          rm swiftshader2.zip
-          curr="$(pwd)/libvk_swiftshader.so"
-          sed -i "s|PATH_TO_CHANGE|$curr|" vk_swiftshader_icd.json
+      run: |
+        wget https://github.com/qarmin/gtk_library_store/releases/download/3.24.0/swiftshader2.zip
+        unzip swiftshader2.zip
+        rm swiftshader2.zip
+        curr="$(pwd)/libvk_swiftshader.so"
+        sed -i "s|PATH_TO_CHANGE|$curr|" vk_swiftshader_icd.json
         
     - name: Test
       shell: bash -l {0}

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -44,7 +44,10 @@ jobs:
     - name: System dependencies
       run: |
         sudo apt update
+        # OpenXR Monado and Yarp
         sudo apt install build-essential git wget unzip cmake ninja-build libeigen3-dev curl patch python3 pkg-config libxcb-dri2-0-dev libxcb-glx0-dev libxcb-icccm4-dev libxcb-keysyms1-dev libx11-dev libx11-xcb-dev libxxf86vm-dev libxrandr-dev libxcb-randr0-dev libvulkan-dev glslang-tools libglvnd-dev libgl1-mesa-dev ca-certificates libusb-1.0-0-dev libudev-dev libhidapi-dev libwayland-dev libuvc-dev libavcodec-dev libopencv-dev libv4l-dev libcjson-dev libsdl2-dev libegl1-mesa-dev mesa-common-dev libace-dev
+        # Additional dependencies
+        sudo apt install libglm-dev libglfw3-dev libglew-dev
         
     - name: Source-based OpenXR installation
       shell: bash -l {0}

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -122,13 +122,23 @@ jobs:
       run: |
         cd build
         make install
+    
+    #Following https://github.com/godotengine/godot/pull/47414    
+    - name: Download SwiftShader
+        if: ${{ matrix.tests }}
+        run: |
+          wget https://github.com/qarmin/gtk_library_store/releases/download/3.24.0/swiftshader2.zip
+          unzip swiftshader2.zip
+          rm swiftshader2.zip
+          curr="$(pwd)/libvk_swiftshader.so"
+          sed -i "s|PATH_TO_CHANGE|$curr|" vk_swiftshader_icd.json
         
     - name: Test
       shell: bash -l {0}
       run: |
-        export DISPLAY=:99
-        sudo Xvfb -ac :99 -screen 0 1280x1024x24 > /dev/null 2>&1 &
         export PATH=$PATH:${GITHUB_WORKSPACE}/install/bin
-        XRT_COMPOSITOR_FORCE_XCB=1 monado-service & 
+        
+        # Run monado-service
+        XRT_COMPOSITOR_FORCE_XCB=1 VK_ICD_FILENAMES=$(pwd)/vk_swiftshader_icd.json DRI_PRIME=0 xvfb-run monado-service & 
         sleep 10
 

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -123,6 +123,14 @@ jobs:
         cd build
         make install
         
+    #Copied from https://github.com/godotengine/godot/pull/47414
+    - name: Download SwiftShader
+      run: |
+        wget https://github.com/godotengine/regression-test-project/releases/download/_ci-deps/swiftshader-ubuntu20.04-20211002.zip
+        unzip swiftshader-ubuntu20.04-20211002.zip
+        curr="$(pwd)/libvk_swiftshader.so"
+        sed -i "s|PATH_TO_CHANGE|$curr|" vk_swiftshader_icd.json
+        
     - name: Test
       shell: bash -l {0}
       run: |
@@ -131,7 +139,6 @@ jobs:
         yarpserver --write &
         
         # Run monado-service
-        XRT_COMPOSITOR_FORCE_XCB=1 XR_RUNTIME_JSON=${GITHUB_WORKSPACE}/install/share/openxr/1/openxr_monado.json xvfb-run ./build/bin/openxrheadset_test
-        sleep 10
+        VK_ICD_FILENAMES=$(pwd)/vk_swiftshader_icd.json DRI_PRIME=0  XRT_COMPOSITOR_FORCE_XCB=1 XR_RUNTIME_JSON=${GITHUB_WORKSPACE}/install/share/openxr/1/openxr_monado.json xvfb-run ./build/bin/openxrheadset_test
     
 

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -65,12 +65,12 @@ jobs:
         sudo apt install wget unzip ninja-build curl patch python3 pkg-config libx11-dev glslang-tools libglvnd-dev ca-certificates libusb-1.0-0-dev libudev-dev libhidapi-dev libwayland-dev libuvc-dev libavcodec-dev libopencv-dev libv4l-dev libcjson-dev libsdl2-dev libegl1-mesa-dev 
         
         #Xvfb, just to have a dummy screen
-        sudo apt install xvfb vulkan-utils
+        sudo apt install xvfb vulkan-utils vulkan-swrast
    
     - name: Vulkan extensions
       shell: bash -l {0}
       run: |
-        VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/lvp_icd.x86_64.json MESA_LOADER_DRIVER_OVERRIDE=zink vulkaninfo
+        LIBGL_ALWAYS_SOFTWARE=1 __GLX_VENDOR_LIBRARY_NAME=mesa VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/lvp_icd.i686.json:/usr/share/vulkan/icd.d/lvp_icd.x86_64.json vulkaninfo
 
     
     - name: Source-based OpenXR installation
@@ -141,6 +141,6 @@ jobs:
         yarpserver --write &
         
         # Run monado-service
-        VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/lvp_icd.x86_64.json MESA_LOADER_DRIVER_OVERRIDE=zink LIBGL_ALWAYS_SOFTWARE=true XRT_COMPOSITOR_FORCE_XCB=1 XR_RUNTIME_JSON=${GITHUB_WORKSPACE}/install/share/openxr/1/openxr_monado.json xvfb-run ./build/bin/openxrheadset_test
+        LIBGL_ALWAYS_SOFTWARE=1 __GLX_VENDOR_LIBRARY_NAME=mesa VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/lvp_icd.i686.json:/usr/share/vulkan/icd.d/lvp_icd.x86_64.json  XRT_COMPOSITOR_FORCE_XCB=1 XR_RUNTIME_JSON=${GITHUB_WORKSPACE}/install/share/openxr/1/openxr_monado.json xvfb-run ./build/bin/openxrheadset_test
     
 

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -78,7 +78,7 @@ jobs:
         cd ${GITHUB_WORKSPACE}
         git clone https://gitlab.freedesktop.org/monado/monado.git --depth 1 --branch ${MONADO_TAG}
         cd monado && mkdir -p build && cd build
-        cmake -G Ninja -DCMAKE_INSTALL_PREFIX=${GITHUB_WORKSPACE}/install -DCMAKE_BUILD_TYPE=Release -DXRT_HAVE_VULKAN:BOOL=OFF -DXRT_FEATURE_SERVICE:BOOL=OFF ..
+        cmake -G Ninja -DCMAKE_INSTALL_PREFIX=${GITHUB_WORKSPACE}/install -DCMAKE_BUILD_TYPE=Release -DXRT_FEATURE_SERVICE:BOOL=OFF ..
         ninja install
         
     - name: Source-based YARP installation

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -70,7 +70,7 @@ jobs:
     - name: Vulkan extensions
       shell: bash -l {0}
       run: |
-        vulkaninfo
+        VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/lvp_icd.x86_64.json MESA_LOADER_DRIVER_OVERRIDE=zink vulkaninfo
 
     
     - name: Source-based OpenXR installation
@@ -141,6 +141,6 @@ jobs:
         yarpserver --write &
         
         # Run monado-service
-        LIBGL_ALWAYS_SOFTWARE=true XRT_COMPOSITOR_FORCE_XCB=1 XR_RUNTIME_JSON=${GITHUB_WORKSPACE}/install/share/openxr/1/openxr_monado.json xvfb-run ./build/bin/openxrheadset_test
+        VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/lvp_icd.x86_64.json MESA_LOADER_DRIVER_OVERRIDE=zink LIBGL_ALWAYS_SOFTWARE=true XRT_COMPOSITOR_FORCE_XCB=1 XR_RUNTIME_JSON=${GITHUB_WORKSPACE}/install/share/openxr/1/openxr_monado.json xvfb-run ./build/bin/openxrheadset_test
     
 

--- a/.github/workflows/conda-forge-ci.yml
+++ b/.github/workflows/conda-forge-ci.yml
@@ -40,7 +40,8 @@ jobs:
       shell: bash -l {0}
       run: |
         # OpenGL is not found on Ubuntu when using conda. Related issue https://github.com/robotology/robotology-superbuild/issues/929
-        mamba install mesa-libgl-cos6-x86_64
+        # See https://github.com/robotology/robotology-superbuild/issues/477
+        mamba install expat-cos6-x86_64 freeglut libselinux-cos6-x86_64 libxau-cos6-x86_64 libxcb-cos6-x86_64 libxdamage-cos6-x86_64 libxext-cos6-x86_64 libxfixes-cos6-x86_64 libxxf86vm-cos6-x86_64 mesa-libgl-cos6-x86_64 mesa-libgl-devel-cos6-x86_64
         
     - name: Configure [Linux]
       if: contains(matrix.os, 'ubuntu')

--- a/.github/workflows/conda-forge-ci.yml
+++ b/.github/workflows/conda-forge-ci.yml
@@ -1,0 +1,59 @@
+name: C++ CI Workflow with conda-forge dependencies
+
+on:
+  push:
+  pull_request:
+  schedule:
+  # * is a special character in YAML so you have to quote this string
+  # Execute a "nightly" build at 2 AM UTC 
+  - cron:  '0 2 * * *'
+
+jobs:
+  build:
+    name: '[${{ matrix.os }}@${{ matrix.build_type }}@conda]'
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        build_type: [Release]
+        os: [ubuntu-latest, windows-2019, macOS-latest]
+      fail-fast: false
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - uses: conda-incubator/setup-miniconda@v2
+      with:
+        mamba-version: "*"
+        channels: conda-forge,robotology,defaults
+        channel-priority: true
+
+    - name: Dependencies
+      shell: bash -l {0}
+      run: |
+        # Compilation related dependencies 
+        mamba install cmake compilers make ninja pkg-config 
+        # Actual dependencies
+        mamba install yarp openxr-sdk
+        
+    - name: Configure [Linux&macOS]
+      if: contains(matrix.os, 'macos') || contains(matrix.os, 'ubuntu')
+      shell: bash -l {0}
+      run: |
+        mkdir -p build
+        cd build
+        cmake -GNinja -DCMAKE_BUILD_TYPE=${{ matrix.build_type }} ..
+      
+    - name: Configure [Windows]
+      if: contains(matrix.os, 'windows')
+      shell: bash -l {0}
+      run: |
+        mkdir -p build
+        cd build
+        cmake -G"Visual Studio 16 2019" -DCMAKE_BUILD_TYPE=${{ matrix.build_type }} ..
+
+    - name: Build 
+      shell: bash -l {0}
+      run: |
+        cd build
+        cmake --build . --config ${{ matrix.build_type }}
+ 

--- a/.github/workflows/conda-forge-ci.yml
+++ b/.github/workflows/conda-forge-ci.yml
@@ -33,7 +33,7 @@ jobs:
         # Compilation related dependencies 
         mamba install cmake compilers make ninja pkg-config 
         # Actual dependencies
-        mamba install glew glfw eigen openxr-sdk yarp
+        mamba install glew glfw eigen glm openxr-sdk yarp
         
     - name: Configure [Linux]
       if: contains(matrix.os, 'ubuntu')

--- a/.github/workflows/conda-forge-ci.yml
+++ b/.github/workflows/conda-forge-ci.yml
@@ -40,7 +40,7 @@ jobs:
       shell: bash -l {0}
       run: |
         # OpenGL is not found on Ubuntu when using conda
-        sudo apt install libx11-dev libx11-xcb-dev libxxf86vm-dev libxrandr-dev libxcb-randr0-dev libvulkan-dev glslang-tools libglvnd-dev libgl1-mesa-dev libegl1-mesa-dev
+        mamba install mesa-libgl-devel-cos7-x86_64
         
     - name: Configure [Linux]
       if: contains(matrix.os, 'ubuntu')

--- a/.github/workflows/conda-forge-ci.yml
+++ b/.github/workflows/conda-forge-ci.yml
@@ -39,8 +39,8 @@ jobs:
       if: contains(matrix.os, 'ubuntu')
       shell: bash -l {0}
       run: |
-        # OpenGL is not found on Ubuntu when using conda
-        mamba install mesa-libgl-devel-cos7-x86_64
+        # OpenGL is not found on Ubuntu when using conda. Related issue https://github.com/robotology/robotology-superbuild/issues/929
+        mamba install mesa-libgl-cos6-x86_64
         
     - name: Configure [Linux]
       if: contains(matrix.os, 'ubuntu')

--- a/.github/workflows/conda-forge-ci.yml
+++ b/.github/workflows/conda-forge-ci.yml
@@ -15,7 +15,7 @@ jobs:
     strategy:
       matrix:
         build_type: [Release]
-        os: [ubuntu-latest, windows-2019, macOS-latest]
+        os: [ubuntu-latest, windows-2019]
       fail-fast: false
 
     steps:
@@ -33,10 +33,10 @@ jobs:
         # Compilation related dependencies 
         mamba install cmake compilers make ninja pkg-config 
         # Actual dependencies
-        mamba install yarp openxr-sdk
+        mamba install glew glfw eigen openxr-sdk yarp
         
-    - name: Configure [Linux&macOS]
-      if: contains(matrix.os, 'macos') || contains(matrix.os, 'ubuntu')
+    - name: Configure [Linux]
+      if: contains(matrix.os, 'ubuntu')
       shell: bash -l {0}
       run: |
         mkdir -p build

--- a/.github/workflows/conda-forge-ci.yml
+++ b/.github/workflows/conda-forge-ci.yml
@@ -39,7 +39,8 @@ jobs:
       if: contains(matrix.os, 'ubuntu')
       shell: bash -l {0}
       run: |
-        mamba install mesalib
+        # OpenGL is not found on Ubuntu when using conda
+        sudo apt install libx11-dev libx11-xcb-dev libxxf86vm-dev libxrandr-dev libxcb-randr0-dev libvulkan-dev glslang-tools libglvnd-dev libgl1-mesa-dev libegl1-mesa-dev
         
     - name: Configure [Linux]
       if: contains(matrix.os, 'ubuntu')

--- a/.github/workflows/conda-forge-ci.yml
+++ b/.github/workflows/conda-forge-ci.yml
@@ -35,6 +35,12 @@ jobs:
         # Actual dependencies
         mamba install glew glfw eigen glm openxr-sdk yarp
         
+    - name: OpenGL [Linux]
+      if: contains(matrix.os, 'ubuntu')
+      shell: bash -l {0}
+      run: |
+        mamba install mesalib
+        
     - name: Configure [Linux]
       if: contains(matrix.os, 'ubuntu')
       shell: bash -l {0}

